fix issue: if bridge Interface existing and the config is not the same with passing config bridge IP(using '--bip='). Docker daemon will failed to start.

moypray opened the pull request:
Signed-off-by: zhangwentao zhangwentao234@huawei.com

moypray committed:
fix issue: if bridge Interface existing and the config is not the same with passing config bridge IP(using '--bip='). Docker daemon will fail to start.

Signed-off-by: Zhang Wentao <zhangwentao234@huawei.com>

moypray commented:
issue link:  https://github.com/docker/docker/issues/10703

phemmer commented:
You can update commits in existing PRs. You can either add additional commits to the branch, or you can even force push to replace the commits.

Also since the PR is created by you, you have the ability to close it as well.

*(In regards to #10847)*

coolljt0725 commented:
You can close #10847

moypray commented:
I have closed #10847 

THX

estesp commented:
Why delete the bridge and start over?  If we want `--bip` to be an overriding flag, then re-assigning the current bridge IP would be a better path, I think.

moypray commented:
@estesp  Thanks for your comments. I don't find the interface to change IP address in netlink package at first. But it is a good idea and i will add one and fix it in that way.

moypray commented:
I have re-fixed this issue by changing the Bridge IP rather than deleting/start over the bridge.
@estesp . ping @jfrazelle

jfrazelle commented:
I'm not comfortable with this, especially us deleting their old bridge

moypray commented:
Now it is not deleting old bridge, but changing the Bridge IP address. 
Yesterday I have submit a patch to change to this way. The original fix is just like you said.
@jfrazelle  thx

moypray commented:
there are two way to change the IP address of a net interface/bridge:
  1. use ioctl to change IPv4 address directly.
  2. use netlink to add a ip and delete old ip ,(just like command "ip addr add/del xxxx/x dev docker0").
Now libcontainer just provide the second way. so i choose the second way to change IP addr for docker0.

estesp commented on daemon/networkdriver/bridge/driver.go:
```diff
@@ -140,13 +141,36 @@ func InitDriver(job *engine.Job) engine.Status {
 			if !networkv4.IP.Equal(bip) {
-				return job.Errorf("bridge ip (%s) does not match existing bridge configuration %s", networkv4.IP, bip)
+				// Configured bridge IP (by --bip="xxxx/x") is not the same with the real bridge IP.
+				// Add a new bridge IP here and remove the old one.
```
I think we should at least add a log entry that the current bridge IP was different than the flag so that there is some "audit" for why/when the bridge device IP changed.  Obviously the passing of the flag should make the user aware that they are requesting it be changed, but it would probably still be good to log.

estesp commented:
Thanks for the updates--I believe we should add a note here: http://docs.docker.com/articles/networking/#building-your-own-bridge that --bip will override current bridge configured IP; seems like that should be the expectation but the docs should be explicit.

I also made a minor code comment about logging this scenario so logs have audit trail of changing the IP for future debug/issue reference.

@LK4D4 any thoughts about this?

moypray commented:
Hi, @estesp , I have added a log entry for that and modified the networking.md file to explicit the  behavior of docker when passing --bip.
Thanks for your comments. Have a nice day~~

estesp commented on docs/sources/articles/networking.md:
```diff
@@ -599,7 +599,8 @@ options are configurable at server startup:
     `docker0` bridge, using standard CIDR notation like
-    `192.168.1.5/24`.
+    `192.168.1.5/24`. If the passing CIDR is not the same with the current
+    bridge IP address, docker will override the old IP with the new CIDR.
```
Please use "If the provided CIDR does not match the current" on the first line.

Second line: ".. , docker will replace the current bridge IP with the provided CIDR."

Thanks!

estesp commented:
ping @LK4D4 - any issues with replacing bridge IP versus failure on `--bip=CIDR` != current bridge IP?  Seems like a better way to handle flag `--bip` being available.

GordonTheTurtle commented:
Can you please sign your commits following these rules:

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

~~~console
$ git clone -b "docker_first" git@github.com:moypray/docker.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

moypray commented:
hi @estesp, done,thanks.

moypray commented:
ping @estesp, 
ping @jfrazelle , how do you feel it ? thx~~
ping @LK4D4

moypray commented:
ping @LK4D4 @jfrazelle 
Obviously, it is a bug of docker. Do you have time to have a look at this? thx~~

estesp commented:
Thanks for your submission, but in discussion with the other maintainers, we would prefer this be an end-user issue to clean up the bridge configuration if they want to change from the current setting.

It would be reasonable to create a different PR to add a bit more documentation about expected failure if `docker0` already has an IP address assigned and a different `--bip` is provided, and even provide the steps to un-associated the current IP and/or delete `docker0`.  Also, the error message could be made a little clearer in this case: "Bridge IP provided does not match the current IP assigned to the bridge: Cannot start daemon"

moypray commented:
@estesp 
Ok,thanks for your explain. I got it although I still think this behavior is more reasonable.

have a nice day :)
